### ISSUE TITLE
Singleton Running, Singleton Queued

### ISSUE THREAD

**Issue by claytondaley** [steinitzu/celery-singleton#14]
As a toy case, assume we're updating a standard deviation field for a specific bucket of values:

 - We add a value to the DB
 - We queue that bucket for calculation
 - The task starts and loads the bucket (and starts work)
 - We add a value to the DB
 - We try to queue the bucket (but singleton blocks it)
 - The task finishes (without incorporating the latest value)

Am I correct in my understanding of how this package works?  It would block the requeue so our value at the end of this sequence isn't fully updated.  

If so, is it hard to adjust this package to only singleton on queued-but-not-running tasks?

**Comment by claytondaley** [comment]
I just realized I *also* need to prevent the long running task from running concurrently.  I assume that makes the requirement much harder since the dequeue also needs to check for a currently running task.

**Comment by claytondaley** [comment]
Closing in favor of https://github.com/cameronmaske/celery-once/issues/114 since more of the infrastructure is already present in that project.

**Comment by steinitzu** [comment]
So I understand correctly?

When task is queued, you want to block duplicate tasks from queue.  
When task is currently running, you want to allow duplicate tasks to be queued.

The way this package works is fairly simple. When you queue the task it creates a lock, when the task is finished running the lock is released and duplicate task can be queued again.

But technically it's possible to release the lock manually at some other time. For instance inside the task itself when it starts, I haven't tried it but  this _might_ work to allow you to queue a new task when the first one starts running:

```python3
@celery.task(bind=True, base=Singleton)
def do_the_thing(self, *args, **kwargs):
    # Release the lock manually now that task is running
    self.release_lock(args, kwargs)	
    # Now run the thing
    ...
```

But there can be timing issues and you can get wrongly blocked occasionally.

Alternatively without celery-singleton, run your worker with concurrency=1 and just delete all pending tasks from the queue before you queue a new one. You might sometimes process the same thing twice, but there's always a tradeoff.

**Comment by claytondaley** [comment]
> Alternatively without celery-singleton, run your worker with concurrency=1 and just delete all pending tasks from the queue before you queue a new one. You might sometimes process the same thing twice, but there's always a tradeoff.

That's really smart!

**Comment by claytondaley** [comment]
Oh wait.  I do need to parallelize the work.  Are you saying use a single worker?  Or would this work with multiple workers?

**Comment by steinitzu** [comment]
This would only work with a single worker that executes one task at a time. My understanding was you wanted to be able to queue max 1 task to run after the currently running task.

You might also want to look into rabbitmq queue length limits.

**Comment by claytondaley** [comment]
Sorry perhaps I wasn't as clear as I needed to be.  I need a singleton for each bucket (max 1 running and 1 queued).

**Comment by steinitzu** [comment]
Then probably the most reliable thing with the least race condition potential is doing this at the broker level in rabbitmq, create a queue for each bucket and set a length limit. Then you should be able to queue another one soon as the first one is acked.

**Comment by claytondaley** [comment]
It was a poor choice to call them "buckets" in my toy case since that carries wrong implications. 
 Here's my actual case:

 - I'm processing messages for patients ("buckets" has some unintended and inaccurate implications)
 - Within a patient I must guaranteed order
 - Between patients order is irrelevant

I need to work patients in parallel.  I certainly *could* bucket them, but there's no reason to think messages will be evenly distributed.  A queue per patient would be amazing but I don't know how I'd pre-subscribe or whatever.  My current thought is to:

 - Access messages in a database (or API endpoint)
 - With each new message put the Patient IDs in the message queue
 - My celery task will request and process all messages for that patient

For obvious reasons, I can only permit 1 active task.  Since messages could queue up after the request, I need to have (up to) a running + a queued task per patient.  This should also let me scale workers without re-bucketing.
